### PR TITLE
PULSEDEV-32102 Add support for large LightGBM datasets

### PR DESCRIPTION
**Pull request by AlbertoEAF**
- Add support to large datasets in LightGBM
  - Use the new LightGBM API: allocate with long parameter (not int)
  - Change the product of integer dimensions of the buffer to a long
  - Copy train data to buffer using a long index
- Replace microsoft's lightgbmlib build by our own
- Update LightGBM to a very recent commit in the master with fast
  prediction support - fc79b36 (LightGBM's #2992)
- Tag this as lightgbmlib=2.3.190 for it is ahead of any available Microsoft lightgbmlib build
- Use the new make-lightgbm which skips builds based on commit id.

### CODE REVIEW

File: openml-lightgbm/lightgbm-builder/make.sh [hunk 1]
```diff
@@ -35,8 +36,8 @@ DIR=lightgbmlib_build
 if [ ! -d "$DIR" ]; then
   echo "Entering the folder."
   cd make-lightgbm
-  echo "Building LightGBM $VERSION"
-  bash make.sh "$VERSION"
+  echo "Building LightGBM $LIGHTGBM_VERSION as lightgbmlib LIGHTGBMLIB_VERSION"
```
Comment (JPDSousa) [inline openml-lightgbm/lightgbm-builder/make.sh:39]: Missing a `$` in the last var

File: openml-lightgbm/openml-lightgbm-provider/pom.xml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                                    <directory>LICENCES</directory>
+                                </resource>
+                            </resources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+
+            <!-- assemble package -->
+            <plugin>
```
Comment (JPDSousa) [inline openml-lightgbm/openml-lightgbm-provider/pom.xml:149]: Out of curiosity: why to you need this? To merge the extra jar into this one?

File: openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public SWIGTrainResources(final int numInstances, final int numFeatures) {
 
         logger.debug("Allocating SWIG train data array.");
         // 1-D Array in row-major-order that stores only the features (excludes label) in double format:
-        this.swigTrainFeaturesDataArray = lightgbmlib.new_doubleArray(numInstances * numFeatures);
+        this.swigTrainFeaturesDataArray = lightgbmlib.new_doubleArray(((long)numInstances) * ((long)numFeatures));
```
Comment (JPDSousa) [inline openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java:96]: Isn't it better to change the parameters of this method to long?

File: openml-lightgbm/lightgbm-builder/make.sh (hunk 1 above)
Comment (AlbertoEAF) [reply openml-lightgbm/lightgbm-builder/make.sh]: Thanks! great catch :) Fixed in the latest commit ammend.

File: openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java (hunk 3 above)
Comment (AlbertoEAF) [reply openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java:96]: I thought the same first, but LightGBM's Java API requires an int for those 2 parameters everywhere.
So it's best not to use a larger data type.

File: openml-lightgbm/openml-lightgbm-provider/pom.xml (hunk 2 above)
Comment (AlbertoEAF) [reply openml-lightgbm/openml-lightgbm-provider/pom.xml:149]: Ideally we should need to carry around only a single jar which has lightgbmlib and the .so's inside.

I think this is needed for that? @JPDSousa 

What I'd like to see is the parent pom assemble a jar with this jar + lightgbmlib so we match the conventions in the openml-feedzai-java for the other providers (as currently the root README for lightgbm is wrong).

File: openml-lightgbm/openml-lightgbm-provider/pom.xml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    <parent>
+        <groupId>com.feedzai</groupId>
+        <artifactId>openml-lightgbm</artifactId>
+        <version>0.0.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.feedzai.openml-lightgbm</groupId>
+    <artifactId>openml-lightgbm-provider</artifactId>
+    <version>1.0.10-ft_af_pulsedev_32102_lgbm_large_datasets_support-SNAPSHOT</version>
```
Comment (shenggwang) [inline openml-lightgbm/openml-lightgbm-provider/pom.xml:23]: Don't forget to change this

File: openml-lightgbm/lightgbm-builder/make.sh [hunk 5]
```diff
@@ -18,13 +18,14 @@
 
 set -e
 
-VERSION="$1"
+LIGHTGBM_VERSION="$1"
+LIGHTGBMLIB_VERSION="$2"
```
Comment (shenggwang) [inline openml-lightgbm/lightgbm-builder/make.sh:22]: So, from [make-lightgbm](https://github.com/feedzai/make-lightgbm/blob/master/make.sh). `LIGHTGBMLIB_VERSION` is the package version. 

But I don't know why is the value of `LIGHTGBM_VERSION` `fc79b366153e8aa5cf2996b4b753e28e5a75e878`? Is it generated?

File: openml-lightgbm/pom.xml [hunk 6]
```diff
@@ -1,213 +1,26 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (shenggwang) [inline openml-lightgbm/pom.xml:1]: Why did you remove this?

File: openml-lightgbm/lightgbm-builder/make.sh (hunk 5 above)
Comment (AlbertoEAF) [reply openml-lightgbm/lightgbm-builder/make.sh:22]: That's the git commit id of LightGBM. make-lightgbm will fetch that instead of the latest release which has the tag v2.3.1.

File: openml-lightgbm/pom.xml (hunk 6 above)
Comment (AlbertoEAF) [reply openml-lightgbm/pom.xml:1]: Thanks Sheng! I had to create a new pom.xml, and forgot to add the license and that. Fixed in the new commit!

File: openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java (hunk 3 above)
Comment (AlbertoEAF) [reply openml-lightgbm/openml-lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java:96]: Basically only our train buffer size can be a long. For LightGBM's datasets, both dimensions are limited to ints.

I don't like the "look" of the code though, is there a more "idiomatic" way to write it? I'm afraid of removing the double parenthesis as it might not cast before the product.

File: openml-lightgbm/lightgbm-builder/make.sh (hunk 5 above)
Comment (shenggwang) [reply openml-lightgbm/lightgbm-builder/make.sh]: If one day they rebased or update all the commits, the hash of the commit might change as well.
If that happens we would need to find the hash again. So, Don't we have other way to get it?

BTW, in that case, we shouldn't name it as `LIGHTGBMLIB_VERSION`, instead `LIGHTGBMLIB_COMMIT_HASH`

Comment (AlbertoEAF) [reply openml-lightgbm/lightgbm-builder/make.sh]: We already do so. But it is the "version" for a lack of a better name and not the hash. Imagine they changed the tag v2.3.1 to point to another commit: the commit changes, but it's still what they call v2.3.1.

If that happens, make-lightgbm checks the hash of that tag and if it doesn't match __commit_id__ in build/ triggers a new build. If it is the same nothing happens.

**Comment by AlbertoEAF** [comment]
@ptraca @JPDSousa regarding producing a fat jar from the parent lightgbm folder, I didn't do it to avoid further complexity. The "provider" `openml-lightgbm` already does it at the end, packaging, lightgbmlib, and the .so's with.

To maintain the conventions of the repo, the provider artifact is `openml-lightgbm` @ `groupId=com.feedzai`.
The full lightgbm implementation is now a `meta-module` containing lightgbm-builder and `openml-lightgbm`. lightgbm-builder and the meta-module are both "hidden" at group `com.feedzai.openml.lightgbm.meta`. I also moved make-lightgbm output artifact to the group above it to be together: `com.feedzai.openml.lightgbm.lightgbmlib`.

File: openml-lightgbm-meta-module/pom.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    <modules>
+        <!-- Builds LightGBM + lightgbmlib.jar -->
+        <module>lightgbm-builder</module>
+        <!-- The provider itself, which requires such artifacts -->
+        <module>openml-lightgbm</module>
+    </modules>
+
+    <properties>
+        <lightgbm.repo.url>https://github.com/microsoft/LightGBM</lightgbm.repo.url>
+        <lightgbm.version>fc79b366153e8aa5cf2996b4b753e28e5a75e878</lightgbm.version>
```
[the remark this answers is not in the document]
Comment (shenggwang) [reply openml-lightgbm-meta-module/pom.xml:35]: I'm just raising some concerns.
But if @JPDSousa says that this is okay, then let's move on.
Just because we are in a democracy 😄

[the remark this answers is not in the document]
Comment (AlbertoEAF) [reply openml-lightgbm-meta-module/pom.xml:35]: This is not the commit of v2.3.1 @shenggwang, this time around is a commit from master because we can't add tags to their repo, and we need something else than their existing tags, but just in the previous release of our provider it was the tag v2.3.1.

File: openml-lightgbm/lightgbm-builder/make.sh (hunk 5 above)
Comment (JPDSousa) [reply openml-lightgbm/lightgbm-builder/make.sh]: Guys, it's really cool the level of critical thinking here, but I guess that we need to establish some ground assumptions here. Feel free to question my statements below:

- Rebases should only be done over development branches. The master branch (as well as hotfix branches) should be monotonic, and are usually marked as protected (i.e. cannot be replayed/rebased). **Referring to commits in protected branches should be safe.**
- Versions should also be immutable: if they are attached to a given commit, then must remain so, otherwise you'll end up with two users, both using version v2.3.1, which are in fact using different versions of the product (i.e. it invalidates the concept of version). **Referring to versions (i.e. git tags) should be safe.**
- I agree with Sheng's suggestion to change the variable name, to make it more consistent 👍

**Comment by AlbertoEAF** [comment]
Fixed :) Ping @shenggwang @JPDSousa for final review.

**Comment by shenggwang** [comment]
The build didn't pass.
It failed on:
```
[INFO] lightgbm-builder ................................... FAILURE [01:11 min]
```
It seemed to me be related with the script:
```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:exec (generate-lightgbm-build) on project lightgbm-builder: Command execution failed. Process exited with an error: 100 (Exit value: 100) -> [Help 1]
```

**Comment by AlbertoEAF** [comment]
> The build didn't pass.
> It failed on:
> 
> ```
> [INFO] lightgbm-builder ................................... FAILURE [01:11 min]
> ```
> 
> It seemed to me be related with the script:
> 
> ```
> [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:exec (generate-lightgbm-build) on project lightgbm-builder: Command execution failed. Process exited with an error: 100 (Exit value: 100) -> [Help 1]
> ```

It failed to fetch things in the network in Ubuntu :sadface:
```
E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/g/glib2.0/libglib2.0-0_2.40.2-0ubuntu1.1+esm3_amd64.deb  HttpError401
E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt-common_0.9.3.5ubuntu3+esm2_all.deb  HttpError401
E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/p/python-apt/python3-apt_0.9.3.5ubuntu3+esm2_amd64.deb  HttpError401
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
The command '/bin/sh -c apt-get update  && apt-get install -y --no-install-recommends         software-properties-common  && apt-add-repository ppa:git-core/ppa  && apt-get update  && apt-get install -y --no-install-recommends         apt-utils         curl         git         jq         libcurl3         libicu52         libunwind8         netcat  && curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash  && apt-get install -y --no-install-recommends git-lfs  && rm -rf /var/lib/apt/lists/*  && rm -rf /etc/apt/sources.list.d/*  echo "UPDATED."' returned a non-zero code: 100
[ERROR] Command execution failed.
```

Maybe I should add this as an image to docker.hub? As in the other day it was a sourceforge downtime which made downloading SWIG fail and the build fail too.

**Comment by JPDSousa** [comment]
> > The build didn't pass.
> > It failed on:
> > ```
> > [INFO] lightgbm-builder ................................... FAILURE [01:11 min]
> > ```
> > 
> > 
> > It seemed to me be related with the script:
> > ```
> > [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:exec (generate-lightgbm-build) on project lightgbm-builder: Command execution failed. Process exited with an error: 100 (Exit value: 100) -> [Help 1]
> > ```
> 
> It failed to fetch things in the network in Ubuntu :sadface:
> 
> ```
> E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/g/glib2.0/libglib2.0-0_2.40.2-0ubuntu1.1+esm3_amd64.deb  HttpError401
> E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt-common_0.9.3.5ubuntu3+esm2_all.deb  HttpError401
> E: Failed to fetch https://esm.ubuntu.com/ubuntu/pool/main/p/python-apt/python3-apt_0.9.3.5ubuntu3+esm2_amd64.deb  HttpError401
> E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
> The command '/bin/sh -c apt-get update  && apt-get install -y --no-install-recommends         software-properties-common  && apt-add-repository ppa:git-core/ppa  && apt-get update  && apt-get install -y --no-install-recommends         apt-utils         curl         git         jq         libcurl3         libicu52         libunwind8         netcat  && curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash  && apt-get install -y --no-install-recommends git-lfs  && rm -rf /var/lib/apt/lists/*  && rm -rf /etc/apt/sources.list.d/*  echo "UPDATED."' returned a non-zero code: 100
> [ERROR] Command execution failed.
> ```
> 
> Maybe I should add this as an image to docker.hub? As in the other day it was a sourceforge downtime which made downloading SWIG fail and the build fail too. Besides, building the image takes 10minutes each time.

Is that a problem in terms of licensing? If not that might be a good improvement. However, let's do it in a separate PR

File: openml-lightgbm/lightgbm-builder/make.sh (hunk 5 above)
Comment (AlbertoEAF) [reply openml-lightgbm/lightgbm-builder/make.sh]: I have no problems choosing another name.

Just to be sure, the request is to rename `LIGHTGBM_VERSION` to `LIGHTGBMLIB_COMMIT_HASH` even if it receives a tag name / version like v2.3.1?

What about `LIGHTGBM_CHECKOUT_TARGET`?
